#### Data Science Capstone Project,  Springboard Bootcamp <br> Title: "Improving Restaurant Reputation Using Yelp User Reviews" <br> Reza Taeb <br> San Francisco, Spring 2018 

## Part 1 - Data Acquiring and Wrangling

In [1]:
# Importing Necessary Packages and Libraries

import pandas as pd 
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl
import matplotlib.cm as cm
from matplotlib.colors import ListedColormap


In [2]:
# Adjust output view

pd.set_option('display.width', 120)
sns.set()

In [3]:
# read CSV files ("review" and "business") 

df_review = pd.read_csv('../review.csv')
df_business = pd.read_csv('../business.csv')

In [4]:
# General Information of data subsets 

print (df_review.info())
print (df_review.head())


print (df_business.info())
print (df_business.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5261669 entries, 0 to 5261668
Data columns (total 9 columns):
review_id      object
user_id        object
business_id    object
stars          int64
date           object
text           object
useful         int64
funny          int64
cool           int64
dtypes: int64(4), object(5)
memory usage: 361.3+ MB
None
                review_id                 user_id             business_id  stars        date  \
0  v0i_UHJMo_hPBq9bxWvW4w  bv2nCi5Qv5vroFiqKGopiw  0W4lkclzZThpx3V65bVgig      5  2016-05-28   
1  vkVSCC7xljjrAI4UGfnKEQ  bv2nCi5Qv5vroFiqKGopiw  AEx2SYEUJmTxVVB18LlCwA      5  2016-05-28   
2  n6QzIUObkYshz4dz2QRJTw  bv2nCi5Qv5vroFiqKGopiw  VR6GpWIda3SfvPC-lg9H3w      5  2016-05-28   
3  MV3CcKScW05u5LVfF6ok0g  bv2nCi5Qv5vroFiqKGopiw  CKC0-MOWMqoeWf6s-szl8g      5  2016-05-28   
4  IXvOzsEMYtiJI0CARmj77Q  bv2nCi5Qv5vroFiqKGopiw  ACFtxLv8pGrrxMm6EgjreA      4  2016-05-28   

                                                text  useful

## Part A  - Data Wrangling on "Business" dataframe 

In [5]:
# Check "business" dataset categories 

cat = df_business['categories']
print (cat.head(5))

0    ['Dentists', 'General Dentistry', 'Health & Me...
1    ['Hair Stylists', 'Hair Salons', "Men's Hair S...
2    ['Departments of Motor Vehicles', 'Public Serv...
3                       ['Sporting Goods', 'Shopping']
4    ['American (New)', 'Nightlife', 'Bars', 'Sandw...
Name: categories, dtype: object


Type of business is specified in ** Categories ** column. Since this project is just focusing on *** Resturants ***, first step is filtering entries that their businesses are "restaurant". 

In [6]:
# Filter "resturant" from original sub dataset "business"

df_restaurant = df_business[df_business['categories'].str.contains('Restaurant')]
df_non_restaurant = df_business[~df_business['categories'].str.contains('Restaurant')]

# Check resturants entires

df_restaurant.info()
df_restaurant.head()

# Check non_restaurant entries

df_non_restaurant.info()
df_non_restaurant.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 54630 entries, 4 to 174558
Data columns (total 15 columns):
business_id     54630 non-null object
name            54630 non-null object
neighborhood    24521 non-null object
address         54311 non-null object
city            54630 non-null object
state           54630 non-null object
postal_code     54529 non-null object
latitude        54630 non-null float64
longitude       54630 non-null float64
stars           54630 non-null float64
review_count    54630 non-null int64
is_open         54630 non-null int64
attributes      54630 non-null object
categories      54630 non-null object
hours           54630 non-null object
dtypes: float64(3), int64(2), object(10)
memory usage: 6.7+ MB
<class 'pandas.core.frame.DataFrame'>
Int64Index: 119937 entries, 0 to 174566
Data columns (total 15 columns):
business_id     119937 non-null object
name            119937 non-null object
neighborhood    43494 non-null object
address         113814 non-nu

business_id                   name neighborhood                         address       city state  \
0  FYWN1wneV18bWNgQjJ2GNg       Dental by Design          NaN        4855 E Warner Rd, Ste B9  Ahwatukee    AZ   
1  He-G7vWjzVUysIKrfNbPUQ    Stephen Szabo Salon          NaN              3101 Washington Rd   McMurray    PA   
2  KQPW8lFf1y5BT2MxiSZ3QA  Western Motor Vehicle          NaN          6025 N 27th Ave, Ste 1    Phoenix    AZ   
3  8DShNS-LuFqpEWIp0HxijA       Sports Authority          NaN  5000 Arizona Mills Cr, Ste 435      Tempe    AZ   
6  kCoE3jvEtg6UVz5SOD3GVw             BDJ Realty    Summerlin        2620 Regatta Dr, Ste 102  Las Vegas    NV   

  postal_code   latitude   longitude  stars  review_count  is_open                                         attributes  \
0       85044  33.330690 -111.978599    4.0            22        1  {'AcceptsInsurance': True, 'ByAppointmentOnly'...   
1       15317  40.291685  -80.104900    3.0            11        1  {'BusinessParking': {'garage': False, 'street'...   
2       85017  33.524903 -112.115310    1.5            18        1                                                 {}   
3       85282  33.383147 -111.964725    3.0             9        0  {'BusinessAcceptsCreditCards': True, 'Restaura...   
6       89128  36.207430 -115.268460    4.0             5        1  {'BusinessAcceptsCreditCards': False, 'ByAppoi...   

                                          categories                                              hours  
0  ['Dentists', 'General Dentistry', 'Health & Me...  {'Friday': '7:30-17:00', 'Tuesday': '7:30-17:0...  
1  ['Hair Stylists', 'Hair Salons', "Men's Hair S...  {'Monday': '9:00-20:00', 'Tuesday': '9:00-20:0...  
2  ['Departments of Motor Vehicles', 'Public Serv...                                                 {}  
3                     ['Sporting Goods', 'Shopping']  {'Monday': '10:00-21:00', 'Tuesday': '10:00-21...  
6  ['Real Estate Services', 'Real Estate', 'Home ...  {'Friday': '8:00-17:00', 'Tuesday': '8:00-17:0...

Initially I divided the original business dataset into “restaurant” and “non_restaurant” subsets.
Glancing to the "categories" column for "non_restaurant" subset, I noticed another similar word to "Restaurant" which is **Food** . There are around 14500 observations that  do not have RESTAURANT but have FOOD as one of their categories tags.  So I need to look at this tag (Food) more carefully. I need to focus more on the entries that have "Food" as one of their categories, since some of them may provide some products or services that are related to the food (such as groceries, etc.) and **Not Serving Food** .    


In [7]:
# Check the entries of the business dataset that do not have "Restaurant" but it has "Food" in their categories. 
# In other words check the entries of "non_restaurant" dataset that have "Food" in their categories

df_food = df_non_restaurant[df_non_restaurant['categories'].str.contains('Food')]
print (df_food.info())
print (df_food.head(5))


print (df_food['categories'].iloc[0:10])
print (df_food['categories'].value_counts())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14449 entries, 8 to 174561
Data columns (total 15 columns):
business_id     14449 non-null object
name            14449 non-null object
neighborhood    6491 non-null object
address         14091 non-null object
city            14449 non-null object
state           14449 non-null object
postal_code     14351 non-null object
latitude        14449 non-null float64
longitude       14449 non-null float64
stars           14449 non-null float64
review_count    14449 non-null int64
is_open         14449 non-null int64
attributes      14449 non-null object
categories      14449 non-null object
hours           14449 non-null object
dtypes: float64(3), int64(2), object(10)
memory usage: 1.8+ MB
None
               business_id                   name     neighborhood                 address      city state  \
8   EsMcGiZaQuG1OOvL9iUFug       Any Given Sundae              NaN   2612 Brandt School Rd   Wexford    PA   
12  xcgFnd-MwkZeO5G2HQ0gAQ  T & 

Looking to the part of the data that have ** Food and not Restaurant**, I figured out that majority of them have "Coffee & Tea" , "Grocery", "Ice Cream & Frozen Yogurt", "Bakeries", "Convenience Store" , and so on. So let's just keep the ones that have Restaurant, since there are enough observations (more than 54 K) and keeping the ones with Food may make it very complex and may bring some biases to our final analysis on reviews (for example the businesses that serve Italian ice cream or american coffee may affect our analysis of whole Italian or American restaurants group). 
Finally, let's exclude the part of the data that has *** Grocery , Convenience Store, Coffee and Tea, Ice Cream and Frozen Yogurt, and Bakeries *** , Because some users might write review about the other parts and not about restuarant. (For example a restaurant that serves Ice Cream or Coffee and Tea may have reviews about these items and not food)

In [8]:
# Filter the df_restaurant from the observations that have other items (like Grocery)

not_restaurant_list = ['Grocery','Bakeries','Coffee & Tea','Ice Cream & Frozen Yogurt','Covenience Store']
not_restaurant_items = '|'.join(not_restaurant_list)
df_restaurant = df_restaurant[df_restaurant['categories'].str.contains(not_restaurant_items) == False]
print (df_restaurant.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50170 entries, 4 to 174558
Data columns (total 15 columns):
business_id     50170 non-null object
name            50170 non-null object
neighborhood    22279 non-null object
address         49870 non-null object
city            50170 non-null object
state           50170 non-null object
postal_code     50073 non-null object
latitude        50170 non-null float64
longitude       50170 non-null float64
stars           50170 non-null float64
review_count    50170 non-null int64
is_open         50170 non-null int64
attributes      50170 non-null object
categories      50170 non-null object
hours           50170 non-null object
dtypes: float64(3), int64(2), object(10)
memory usage: 6.1+ MB
None


In [9]:
# Check locations of businesses ("city" and "state" columns)

print (df_restaurant['state'].value_counts())
print (df_restaurant['city'].value_counts())

# Check city='stuatgart' and state='01' as examples to figure out whether the language of reviews of business in all cities are English or not

df_stuatgart = df_restaurant[df_restaurant['city'] == 'Stuttgart']
print (df_stuatgart.head(5))

df_01 = df_restaurant[df_restaurant['state']=='01']
print (df_01.head(5))

# Check a business's reviews (A restaurant in Stuatgart )

df_example_review = df_review[df_review['business_id'] == 'o9eMRCWt5PkpLDE0gOPtcQ']
print (df_example_review)
print (df_example_review['text'].iloc[:4])

ON     12291
AZ      9802
NV      6680
QC      4528
OH      4521
NC      3622
PA      3396
BW      1558
WI      1387
EDH     1210
IL       577
SC       198
MLN       89
CHE       59
HLD       59
NYK       51
FIF       31
WLN       20
ELN       19
C         17
NY         9
01         9
NI         8
ST         4
ESX        3
IN         3
BY         3
CO         2
VS         2
WHT        1
CA         1
30         1
3          1
AK         1
HU         1
ZET        1
B          1
VA         1
PKN        1
GLG        1
6          1
Name: state, dtype: int64
Toronto                                 6424
Las Vegas                               5533
Phoenix                                 3361
Montréal                                2981
Charlotte                               2309
Pittsburgh                              2039
Scottsdale                              1334
Cleveland                               1278
Edinburgh                               1246
Mississauga                         

By checking "state" column of "business" dataset, I figured out businesses' locations are diverse (including non English spoken countries). So I checked the *** text *** of businesses ("review" subdataset) to figure out whether the language of reviews are English or not. It was seen that the reviews' texts are not in English so I have to filter the original dataset from the English-spoken countries by using 'state' column and checking their addresses.    
I made a new subdata set, ** df_restaurant_eng **

In [10]:
# Filter "Non_English" reviews using 'state' column

no_eng_list = ['BW','C','01','NI','ST','6','30','3','BY','B','WHT','CA','HU']
df_restaurant_eng = df_restaurant[~df_restaurant.state.isin(no_eng_list)]

print (df_restaurant_eng.head())
print (df_restaurant_eng.info())

df_restaurant_eng['state'].value_counts()

               business_id                        name        neighborhood                  address            city  \
4   PfOCPjBrlQAnz__NXj9h_w    Brick House Tavern + Tap                 NaN             581 Howe Ave  Cuyahoga Falls   
14  fNMVV_ZX7CJSDWQGdOM8Nw  Showmars Government Center              Uptown             600 E 4th St       Charlotte   
15  l09JfMeQ6ynYs5MCJtrcmQ              Alize Catering  Yonge and Eglinton            2459 Yonge St         Toronto   
25  1K4qrnfyzKzGgJPBEcJaNQ      Chula Taberna Mexicana         Leslieville    1058 Gerrard Street E         Toronto   
28  Dj0S-Oe4ytRJzMGUPgYUkw                Panera Bread                 NaN  38295 Chestnut Ridge Rd          Elyria   

   state postal_code   latitude  longitude  stars  review_count  is_open  \
4     OH       44221  41.119535 -81.475690    3.5           116        1   
14    NC       28202  35.221647 -80.839345    3.5             7        1   
15    ON     M4P 2H6  43.711399 -79.399339    3.0        

ON     12291
AZ      9802
NV      6680
QC      4528
OH      4521
NC      3622
PA      3396
WI      1387
EDH     1210
IL       577
SC       198
MLN       89
HLD       59
CHE       59
NYK       51
FIF       31
WLN       20
ELN       19
NY         9
IN         3
ESX        3
CO         2
VS         2
ZET        1
GLG        1
AK         1
VA         1
PKN        1
Name: state, dtype: int64

Using "state" column, I added a new column, "country", to **** df_restaurant_eng **** dataset for future usage. I am thinking maybe restaurants' customers in different English Spoken countries - US, Britain, Canada - show different behaviors and have different expectations.  

In [11]:
# Add 'country' Column to english spoken restaurant dataset 

df_restaurant_eng.loc[:,'country']=np.nan

def countryfinder(x):
    if x in ['ON','QC']:
        return 'CAN' 
    elif x in ['AZ','NV','OH','NC','PA','WI','IL','SC','NY','IN','VA']:
        return 'USA'
    else: 
        return 'UK'

df_restaurant_eng.loc[:,'country'] = df_restaurant_eng.loc[:,'state'].apply(lambda x:countryfinder(x))

print (df_restaurant_eng.head(5))
print (df_restaurant_eng['country'].value_counts())

               business_id                        name        neighborhood                  address            city  \
4   PfOCPjBrlQAnz__NXj9h_w    Brick House Tavern + Tap                 NaN             581 Howe Ave  Cuyahoga Falls   
14  fNMVV_ZX7CJSDWQGdOM8Nw  Showmars Government Center              Uptown             600 E 4th St       Charlotte   
15  l09JfMeQ6ynYs5MCJtrcmQ              Alize Catering  Yonge and Eglinton            2459 Yonge St         Toronto   
25  1K4qrnfyzKzGgJPBEcJaNQ      Chula Taberna Mexicana         Leslieville    1058 Gerrard Street E         Toronto   
28  Dj0S-Oe4ytRJzMGUPgYUkw                Panera Bread                 NaN  38295 Chestnut Ridge Rd          Elyria   

   state postal_code   latitude  longitude  stars  review_count  is_open  \
4     OH       44221  41.119535 -81.475690    3.5           116        1   
14    NC       28202  35.221647 -80.839345    3.5             7        1   
15    ON     M4P 2H6  43.711399 -79.399339    3.0        

/Users/rtmac/anaconda3/lib/python3.6/site-packages/pandas/core/indexing.py:337: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[key] = _infer_fill_value(value)
/Users/rtmac/anaconda3/lib/python3.6/site-packages/pandas/core/indexing.py:517: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


Now, let's look deeper into "categories" column to check type of foods in resturants. I want to figure out which types of food exist more in the dataset, so I may bring type of food as a factor in my analysis. 

In [12]:
# Check resturant 'categories' in order to find most popular types of food 

print (df_restaurant_eng['categories'].value_counts())

['Restaurants', 'Pizza']                                                                                                                                                                                                 983
['Pizza', 'Restaurants']                                                                                                                                                                                                 982
['Mexican', 'Restaurants']                                                                                                                                                                                               904
['Restaurants', 'Mexican']                                                                                                                                                                                               858
['Restaurants', 'Chinese']                                                                                          

By just taking a glance into 'categories' column, some of the very popular types of food in the dataset are:   
**** American, Italian, Chinese, Mexican, Indian, Japanese, Thai ****.   
I added the new column *** 'food_type' *** for future refrence in my analysis. 

In [13]:
# Add a new column 'food_type' to 'df_restaurant_eng' dataset


def typefinder(x):
    if ('Italian' in x) and ('American' not in x) and ('Japanese' not in x) and  ('Indian' not in x) and ('Mexican' not in x) and ('Thai' not in x) and ('Chinese' not in x):
        return 'Italian'
    if ('American' in x) and ('Italian' not in x) and ('Japanese' not in x) and  ('Indian' not in x) and ('Mexican' not in x) and ('Thai' not in x) and ('Chinese' not in x): 
        return 'American'
    if ('Mexican' in x) and ('American' not in x) and ('Japanese' not in x) and  ('Indian' not in x) and ('Italian' not in x) and ('Thai' not in x) and ('Chinese' not in x):
        return 'Mexican'
    if ('Indian' in x) and ('American' not in x) and ('Japanese' not in x) and  ('Italian' not in x) and ('Mexican' not in x) and ('Thai' not in x) and ('Chinese' not in x):
        return 'Indian'
    if ('Chinese' in x) and ('American' not in x) and ('Japanese' not in x) and  ('Indian' not in x) and ('Mexican' not in x) and ('Thai' not in x) and ('Italian' not in x):
        return 'Chinese'
    if ('Thai' in x) and ('American' not in x) and ('Japanese' not in x) and  ('Indian' not in x) and ('Mexican' not in x) and ('Italian' not in x) and ('Chinese' not in x):
        return 'Thai'
    if ('Japanese' in x) and ('American' not in x) and ('Italian' not in x) and  ('Indian' not in x) and ('Mexican' not in x) and ('Thai' not in x) and ('Chinese' not in x):
        return 'Japanese'
    else:
        return None


df_restaurant_eng['food_type'] = df_restaurant_eng['categories'].apply(lambda x:typefinder(x))

print (df_restaurant_eng['food_type'].head(5))
print (df_restaurant_eng['food_type'].value_counts())
print (df_restaurant_eng['food_type'].count())

print (df_restaurant_eng.info())

4     American
14    American
15     Italian
25     Mexican
28        None
Name: food_type, dtype: object
American    8401
Italian     3788
Mexican     3604
Chinese     3318
Japanese    1972
Indian      1209
Thai         989
Name: food_type, dtype: int64
23281
<class 'pandas.core.frame.DataFrame'>
Int64Index: 48564 entries, 4 to 174558
Data columns (total 17 columns):
business_id     48564 non-null object
name            48564 non-null object
neighborhood    22075 non-null object
address         48266 non-null object
city            48564 non-null object
state           48564 non-null object
postal_code     48470 non-null object
latitude        48564 non-null float64
longitude       48564 non-null float64
stars           48564 non-null float64
review_count    48564 non-null int64
is_open         48564 non-null int64
attributes      48564 non-null object
categories      48564 non-null object
hours           48564 non-null object
country         48564 non-null object
food_type       2328

/Users/rtmac/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:23: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


In [14]:
# Number of restaurant (count) and number of reviews (sum) for each type of food and star

print (df_restaurant_eng.groupby(['food_type','stars'])['review_count'].agg(['count','mean','sum']))

                 count        mean     sum
food_type stars                           
American  1.0       21    4.952381     104
          1.5       99   15.292929    1514
          2.0      398   26.060302   10372
          2.5      965   48.964767   47251
          3.0     1821   61.663921  112290
          3.5     2273   92.252970  209691
          4.0     1970  160.464975  316116
          4.5      717  137.789400   98795
          5.0      137   20.087591    2752
Chinese   1.0       19    6.421053     122
          1.5       41    8.634146     354
          2.0      188   20.265957    3810
          2.5      410   29.087805   11926
          3.0      852   35.836854   30533
          3.5      962   59.067568   56823
          4.0      644   64.026398   41233
          4.5      172   46.046512    7920
          5.0       30    6.533333     196
Indian    1.0        2    3.000000       6
          1.5       14   10.642857     149
          2.0       27   11.407407     308
          2

In [15]:
# Information of final dataset (Restaurant that serve one of the 7 most popular food in English Spoken Countries)

print (df_restaurant_eng[df_restaurant_eng['food_type'].notnull()].info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 23281 entries, 4 to 174558
Data columns (total 17 columns):
business_id     23281 non-null object
name            23281 non-null object
neighborhood    9786 non-null object
address         23196 non-null object
city            23281 non-null object
state           23281 non-null object
postal_code     23256 non-null object
latitude        23281 non-null float64
longitude       23281 non-null float64
stars           23281 non-null float64
review_count    23281 non-null int64
is_open         23281 non-null int64
attributes      23281 non-null object
categories      23281 non-null object
hours           23281 non-null object
country         23281 non-null object
food_type       23281 non-null object
dtypes: float64(3), int64(2), object(12)
memory usage: 3.2+ MB
None


## Part B  - Data Wrangling on "Review" dataframe 

In [16]:
# Get the unique business_ids of 'df_review' and 'df_restaurant_eng' datasets

df_review_business_list = df_review['business_id'].unique()
df_restaurant_eng_business_list = df_restaurant_eng['business_id'].unique()

#print (type(df_review_business_list))
print ("Number of unique business ids in review dataframe:",len(df_review_business_list))


#print (type(df_restaurant_eng_business_list))
print ("Number of unique business ids in restaurant_eng dataframe:",len(df_restaurant_eng_business_list))


# Find the matching list of business_ids between two datasets ("restaurant" and "review")

business_id_match = set(df_restaurant_eng['business_id']) & set(df_review['business_id'])
#print (business_id_match)
print ("The number of matching business ids in two dataframes:",len(business_id_match))

Number of unique business ids in review dataframe: 174567
Number of unique business ids in restaurant_eng dataframe: 48564
The number of matching business ids in two dataframes: 48564


So based on the matching number (48564), I could say all of the entries in "df_restaurant_eng" have some reviews in "df_review". But some of the entries in "df_review" do not have any matching row in "df_restaurant_eng".

In [17]:
# Make "df_review_restaurant_eng" for restaurants in english spoken countries 

# Since the original df_review is big (more than 5 Million rows), I had to made it into few slices and do filtering 
# And then merge the datasets 

business_id_match_list = list(business_id_match)

# 1 

df_review_1 = df_review.iloc[:1000000,]
df_review_restaurant_eng1 = df_review_1[df_review_1['business_id'].isin(business_id_match_list)]
print (df_review_restaurant_eng1.info())

# 2

df_review_2 = df_review.iloc[1000000:2000000,]
df_review_restaurant_eng2 = df_review_2[df_review_2['business_id'].isin(business_id_match_list)]
print (df_review_restaurant_eng2.info())


# 3 

df_review_3 = df_review.iloc[2000000:3000000,]
df_review_restaurant_eng3 = df_review_3[df_review_3['business_id'].isin(business_id_match_list)]
print (df_review_restaurant_eng3.info())


# 4 

df_review_4 = df_review.iloc[3000000:4000000,]
df_review_restaurant_eng4 = df_review_4[df_review_4['business_id'].isin(business_id_match_list)]
print (df_review_restaurant_eng4.info())

# 5 

df_review_5 = df_review.iloc[4000000:5000000,]
df_review_restaurant_eng5 = df_review_5[df_review_5['business_id'].isin(business_id_match_list)]
print (df_review_restaurant_eng5.info())

# 6 

df_review_6 = df_review.iloc[5000000:,]
df_review_restaurant_eng6 = df_review_6[df_review_6['business_id'].isin(business_id_match_list)]
print (df_review_restaurant_eng6.info())
print (type(df_review_restaurant_eng6))

<class 'pandas.core.frame.DataFrame'>
Int64Index: 562587 entries, 0 to 999994
Data columns (total 9 columns):
review_id      562587 non-null object
user_id        562587 non-null object
business_id    562587 non-null object
stars          562587 non-null int64
date           562587 non-null object
text           562587 non-null object
useful         562587 non-null int64
funny          562587 non-null int64
cool           562587 non-null int64
dtypes: int64(4), object(5)
memory usage: 42.9+ MB
None
<class 'pandas.core.frame.DataFrame'>
Int64Index: 562752 entries, 1000003 to 1999998
Data columns (total 9 columns):
review_id      562752 non-null object
user_id        562752 non-null object
business_id    562752 non-null object
stars          562752 non-null int64
date           562752 non-null object
text           562752 non-null object
useful         562752 non-null int64
funny          562752 non-null int64
cool           562752 non-null int64
dtypes: int64(4), object(5)
memory usage:

In [18]:
# Now it is time to merge these 6 subsets ! 

dataframes = [df_review_restaurant_eng1, df_review_restaurant_eng2, df_review_restaurant_eng3, df_review_restaurant_eng4, df_review_restaurant_eng5, df_review_restaurant_eng6]

df_review_restaurant_eng = pd.concat(dataframes)
print (df_review_restaurant_eng.info())

print (df_review_restaurant_eng['business_id'].value_counts())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2957735 entries, 0 to 5261668
Data columns (total 9 columns):
review_id      object
user_id        object
business_id    object
stars          int64
date           object
text           object
useful         int64
funny          int64
cool           int64
dtypes: int64(4), object(5)
memory usage: 225.7+ MB
None
4JNXUYY8wbaaDmk3BPzlWw    7362
RESDUcs7fIiihp38-d6_6g    7006
K7lWdNUhCbcnEvI0NhGewg    5951
cYwJA2A6I12KNkm2rtXd5g    5448
DkYS3arLOhA8si5uUEmHOw    4870
f4x1YBxkLrZg652xt2KR5g    4774
2weQS-RnoOBhb1KsHKyoSQ    4018
KskYqH1Bi7Z_61pH6Om8pg    3965
eoHdUeQDNgQ6WYEnP2aiRw    3911
ujHiaprwCQ5ewziu0Vi9rw    3836
iCQpiavjjPzJ5_3gPD5Ebg    3741
El4FC8jcawUVgw_0EIcbaQ    3445
hihud--QRriCYZw1zZvW4g    3264
7sPNbCx7vGAaH7SbNPZ6oA    3064
g8OnV26ywJlZpezdBnOWUQ    3051
XZbuPXdyA0ZtTu3AzqtQhg    3011
OETh78qcgDltvHULowwhJg    2933
P7pxQFqr7yBKMMI2J51udw    2771
XXW_OFaYQkkGOGniujZFHg    2700
HhVmDybpU7L50Kb5A0jXTg    2673
yfxDa8RFOvJPQh0rN

In [19]:
# Initial check of "df_review_restaurant_eng" 


print (df_review_restaurant_eng.info())
print (df_review_restaurant_eng.head(5))

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2957735 entries, 0 to 5261668
Data columns (total 9 columns):
review_id      object
user_id        object
business_id    object
stars          int64
date           object
text           object
useful         int64
funny          int64
cool           int64
dtypes: int64(4), object(5)
memory usage: 225.7+ MB
None
                review_id                 user_id             business_id  stars        date  \
0  v0i_UHJMo_hPBq9bxWvW4w  bv2nCi5Qv5vroFiqKGopiw  0W4lkclzZThpx3V65bVgig      5  2016-05-28   
1  vkVSCC7xljjrAI4UGfnKEQ  bv2nCi5Qv5vroFiqKGopiw  AEx2SYEUJmTxVVB18LlCwA      5  2016-05-28   
2  n6QzIUObkYshz4dz2QRJTw  bv2nCi5Qv5vroFiqKGopiw  VR6GpWIda3SfvPC-lg9H3w      5  2016-05-28   
3  MV3CcKScW05u5LVfF6ok0g  bv2nCi5Qv5vroFiqKGopiw  CKC0-MOWMqoeWf6s-szl8g      5  2016-05-28   
4  IXvOzsEMYtiJI0CARmj77Q  bv2nCi5Qv5vroFiqKGopiw  ACFtxLv8pGrrxMm6EgjreA      4  2016-05-28   

                                                text  useful

#### Now it is time to save the cleaned and modified data frames (restaurant.csv & restaurant_eng.csv & review) into the main folder for further use in next steps ! 

In [20]:
# Save the cleaned and modified datasets (df_restaurant , df_restaurant_eng , and df_review_restaurant_eng) into the main folder for the next steps ! 

df_restaurant.to_csv("../restaurant.csv", index=False , encoding = 'utf8')
df_restaurant_eng.to_csv("../restaurant_eng.csv" , index=False , encoding = 'utf8')
df_review_restaurant_eng.to_csv("../review_restaurant_eng.csv", index=False , encoding = 'utf8')